### ISSUE TITLE
User should be able to specify group URL at creation

### ISSUE THREAD

**Issue by boonebgorges** [cuny-academic-commons/openlab-theme#141]
See #138 and https://3.basecamp.com/3165813/buckets/1307067/uploads/960825138

Details to follow.

**Comment by bree-z** [comment]
This has been thoroughly tested, but I'll leave open in case @jennaspevack still wants to take a look.

**Comment by jennaspevack** [comment]
@bree-z Just curious if you tested this functionality when cloning. I tried cloning and noticed that both the green check and the error message will appear if you type in a url with errors, edit it, and then click on another field. It doesn't allow you submit, but keeps the green check icon. Do you mind double checking that?

**Comment by bree-z** [comment]
@jennaspevack Thanks, Jenna!  We did test this, but I think we were focused on functionality and totally missed the green check icon.  :-/  It looks like it functions this way for any kind of group creation.

Boone, here's a screenshot:

![urlerror](https://user-images.githubusercontent.com/7189824/39050944-331ae56a-4475-11e8-88e4-5c2cccb6f745.png)

**Comment by boonebgorges** [comment]
Thanks for the report and the screenshots. I originally wrote the checkbox stuff to reflect only the AJAX uniqueness check. I've refactored so that it now reflects both the uniqueness and the validity checks. It's still not 100% perfect because the AJAX check doesn't happen instantaneously, but it feels pretty good in my testing.

**Comment by bree-z** [comment]
Thanks, Boone!  This looks good to me.  I wasn't able to enter an incorrect URL without getting the  red icon and error message.

@jennaspevack I'll leave this open in case you want to take another look.

**Comment by bree-z** [comment]
Hi Boone,

@charliemedwards  reported the following issue:

Got some weird behavior around the group URL - flagged it as red when there’s absolutely no way anyone had done Victorian Novels in Context before! :) When I navigated to the URL field and added “-” it displayed green checkbox; when I backed up and erased the “-” it stayed green; I was also able to blank it out and it still stayed green

I tried to recreate, and was able to reproduce in one test but not others:

- First, I tried a few unique seminar URLs, and wasn’t able to reproduce.
- Then, I tried a few unique but similar to other group URLs and was able to reproduce.
- I created a project with the name "Hiking in the Rhodopes with a 3 year old" (similar to the existing project "Hiking in the Rhodopes"), which generated the URL “hiking-in-the-rhodopes-with-a-3-year-old”, and a red field with the error icon.
- I took a screenshot of this, and after I took the screenshot, the field changed and displayed the green check mark.  I’m not sure how long it was (15-20 seconds?) but long enough to be confusing.  
- I cleared everything and tried again, with the same name/URL.  This time the field didn’t turn green.  I left it for a few mintues and still nothing changed.  Then, I changed the group name to “Hiking in the Rhodopes with a 3 year old kid”.  The URL didn’t change (it was still hiking-in-the-rhodopes-with-a-3-year-old), but I got the green check.  
- I tried another group name that was unique but similar to another, but couldn’t reproduce.  New Group: Chokladbollar Fan Club Yum Yum, similar to: Chokladbollar Fan Club.  The URL field for this one displayed the green check immediately.  
- I wasn’t able to reproduce with any other groups.  

- I was also able to reproduce the last part of what Charlie describes above.  
- If I delete a URL from the field, I still see the check mark even though it’s blank.  Trying to submit without the group URL works as expected and I get an error that I have to fill in the empty field.  

<img width="1208" alt="groupurlred" src="https://user-images.githubusercontent.com/7189824/39439473-b5ee8232-4c75-11e8-9386-f283d166668a.png">

<img width="892" alt="groupurlchangedname" src="https://user-images.githubusercontent.com/7189824/39439480-bb725b48-4c75-11e8-8a4b-801aea2ee801.png">

<img width="893" alt="groupurlblank" src="https://user-images.githubusercontent.com/7189824/39439490-c2c6c42e-4c75-11e8-8963-e53ea0642471.png">

**Comment by boonebgorges** [comment]
Regarding the '-' issue: I've added some more validation of correct URL formatting. The error now thrown looks like the following screenshot:

![screenshot_2018-04-30_13-06-02](https://user-images.githubusercontent.com/246627/39442557-3f6b7500-4c77-11e8-8196-3ac186fe1f47.png)

Regarding the other near-duplicate issue: I can't reproduce. My guess is that the dupe-check is not being triggered. Because it's AJAX-powered (and thus fairly slow), it only runs on blur, which is to say when the URL field loses focus. This can happen when you tab or click away, or when the window loses focus. I'm guessing that's what happened in your "15-20 second" case.

If you are still seeing issues with this guidance in mind, please open your browser's Network console (F12, then find and click the Network tab), and filter by XHR requests. Then, start playing with the near-duplicate cases. AJAX requests should be fired on blur. If they are *not* being fired when you'd expect, let me know. If they *are*, but are incorrectly showing that the URL is a duplicate, click into the AJAX request in the Network tab and tell me what's on the Headers and Preview or Response tabs.

![screenshot_2018-04-30_13-08-37](https://user-images.githubusercontent.com/246627/39442698-c601bc3c-4c77-11e8-8572-f2ab9b576c69.png)
![screenshot_2018-04-30_13-09-00](https://user-images.githubusercontent.com/246627/39442704-c8f975ec-4c77-11e8-85b8-b882b888950e.png)
![screenshot_2018-04-30_13-08-51](https://user-images.githubusercontent.com/246627/39442706-c91a19fa-4c77-11e8-857a-cf767e97ff14.png)

**Comment by bree-z** [comment]
Thanks, Boone!

I might have found a new pattern.  When I tab to the URL field it works as expected, but if I click with my mouse I get an incorrect duplicate URL message.  It doesn't appear to be related to near duplicate URLs -- this happened with totally unique URLs.  

When I clicked away from the URL field the error remained, and when I switched to another tab and came back, it remained.  

Here's what was in the Headers tab of my browser:

**General**
Request URL: http://openlabdev.commonsinabox.org/wp-admin/admin-ajax.php?action=openlab_group_url_validate&groupId=0&url=this-name-has-never-been-used-for-a-seminar-before
Request Method: GET
Status Code: 200 OK
Remote Address: 146.96.128.253:80
Referrer Policy: no-referrer-when-downgrade

**Response Headers**
Cache-Control: no-cache, must-revalidate, max-age=0
Connection: Keep-Alive
Content-Length: 16
Content-Type: application/json; charset=UTF-8
Date: Mon, 30 Apr 2018 21:00:00 GMT
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Keep-Alive: timeout=5, max=100
Referrer-Policy: strict-origin-when-cross-origin
Server: Apache
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Robots-Tag: noindex

**Request Headers**
Accept: */*
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.9
Connection: keep-alive
Cookie: wp-settings-time-8=1522867153; __utma=162204206.1901421504.1512743872.1522422972.1524606018.4; __utmz=162204206.1524606018.4.2.utmcsr=openlabdev.commonsinabox.org|utmccn=(referral)|utmcmd=referral|utmcct=/; wordpress_test_cookie=WP+Cookie+check; bp-activity-oldestpage=1; wordpress_c2f1c3a0bca2cd88b5f65da3ff73562f=testfaculty%7C1525294437%7CC1qVB4dbBX79WhOd14K24fdjsD9sahUYVjBZOugTVrp%7C23420b82f0cae01321611a6a569f13d0ec8241636d9aac448f6b96b5a92839a6; wordpress_logged_in_c2f1c3a0bca2cd88b5f65da3ff73562f=testfaculty%7C1525294437%7CC1qVB4dbBX79WhOd14K24fdjsD9sahUYVjBZOugTVrp%7C3330516642d6bec6f1fd585076963ee4fab51c66eefeb6e4e603c0f73496c266
Host: openlabdev.commonsinabox.org
Referer: http://openlabdev.commonsinabox.org/groups/create/step/group-details/?group_type=course&new=true
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/66.0.3359.139 Safari/537.36
X-Requested-With: XMLHttpRequest
action: openlab_group_url_validate
groupId: 0
url: this-name-has-never-been-used-for-a-seminar-before

**Preview tab:**

{success: true}
success: true

![urlerroraftermouseclick](https://user-images.githubusercontent.com/7189824/39450433-5256a802-4c99-11e8-836a-fab4296528aa.png)
